SU2_CFD/src/solvers/CNEMONSSolver.cpp

```diff
@@ -804,99 +792,137 @@ void CNEMONSSolver::BC_IsothermalCatalytic_Wall(CGeometry *geometry,
-          for (jSpecies = 0; jSpecies < nSpecies; jSpecies++) {
-            Jacobian_j[nSpecies+nDim][iSpecies] += Jacobian_j[jSpecies][iSpecies]*hs[iSpecies];
+          // Vib.-El. Temperature
+          for (iSpecies = 0; iSpecies < nSpecies; iSpecies++) {
+            for (jSpecies = 0; jSpecies < nSpecies; jSpecies++)
+              Jacobian_j[nSpecies+nDim+1][iSpecies] += Jacobian_j[jSpecies][iSpecies]*eves[iSpecies];
+            Jacobian_j[nSpecies+nDim+1][nSpecies+nDim+1] += Res_Visc[iSpecies]/Area*Cvve[iSpecies];
           }
-          Jacobian_j[nSpecies+nDim][nSpecies+nDim] += Res_Visc[iSpecies]/Area*(Ru/Ms[iSpecies] +
-                                                                               Cvtrs[iSpecies]  );
-          Jacobian_j[nSpecies+nDim][nSpecies+nDim+1] += Res_Visc[iSpecies]/Area*Cvve[iSpecies];
+
+          /*--- Multiply by the transformation matrix and store in Jac. ii ---*/
+          for (iVar = 0; iVar < nVar; iVar++)
+            for (jVar = 0; jVar < nVar; jVar++)
+              for (kVar = 0; kVar < nVar; kVar++)
+                Jacobian_i[iVar][jVar] += Jacobian_j[iVar][kVar]*dVdU[kVar][jVar]*Area;
+
+          /*--- Apply to the linear system ---*/
+          Jacobian.SubtractBlock(iPoint, iPoint, Jacobian_i);
         }
 
-        // Vib.-El. Temperature
-        for (iSpecies = 0; iSpecies < nSpecies; iSpecies++) {
-          for (jSpecies = 0; jSpecies < nSpecies; jSpecies++)
-            Jacobian_j[nSpecies+nDim+1][iSpecies] += Jacobian_j[jSpecies][iSpecies]*eves[iSpecies];
-          Jacobian_j[nSpecies+nDim+1][nSpecies+nDim+1] += Res_Visc[iSpecies]/Area*Cvve[iSpecies];
+      } else {
+
+        if (implicit) {
+          SU2_MPI::Error("Implicit not currently available for partially catalytic wall.",CURRENT_FUNCTION);
         }
 
-        /*--- Multiply by the transformation matrix and store in Jac. ii ---*/
-        for (iVar = 0; iVar < nVar; iVar++)
-          for (jVar = 0; jVar < nVar; jVar++)
-            for (kVar = 0; kVar < nVar; kVar++)
-              Jacobian_i[iVar][jVar] += Jacobian_j[iVar][kVar]*dVdU[kVar][jVar]*Area;
+        /*--- Identify the boundary ---*/
+        string Marker_Tag = config->GetMarker_All_TagBound(val_marker);
+
+        /*--- Get isothermal wall temp ----*/
+        su2double Tw = config->GetIsothermal_Temperature(Marker_Tag);
+
+        /*--- Get wall catalytic efficiency ----*/
+        su2double gam = config->GetCatalytic_Efficiency();
+
+        /*--- Get gas model ---*/
+        string gas_model = config->GetGasModel();
+
+        if (gas_model == string("N2")) {
+          Res_Visc[0] = gam*Vi[RHOS_INDEX+1]/Vi[RHO_INDEX]*rho*sqrt(RuSI*Tw/2/Ms[1]/PI_NUMBER)*Area;
+          Res_Visc[1] = -gam*Vi[RHOS_INDEX+1]/Vi[RHO_INDEX]*rho*sqrt(RuSI*Tw/2/Ms[1]/PI_NUMBER)*Area;
+        } else if (gas_model == string("AIR-5")) {
+          Res_Visc[0] = gam*Vi[RHOS_INDEX]/Vi[RHO_INDEX]*rho*sqrt(RuSI*Tw/2/Ms[0]/PI_NUMBER)*Area;
```

Comment:
I think we can have this handled in fluidmodel.  It seems theres a lot of repetition. Maybe someone will have a clever idea.

Reply:
quite... at least factor out the common parts. also there is probably something more expressive than RHOS_INDEX+1? or is that simply the first species?

Reply:
Yeah that's simply the first species. One thought I had is to use a for-loop with a vector encoding complementary pairs for this residual, that could be stored in fluid model, or read in from the config.

Reply:
It may be work creating a function it self for this, similiar to the GetChemistry() function.